ManageIQ/manageiq-ui-classic
# Fix namespaced engine webpack's pack location

Introduced in https://github.com/ManageIQ/manageiq-ui-classic/pull/1919

A manageiq plugin's name can't reliably be determined from a gem's path, specifically when installed via a git gem, because the path ends up being something like:

```
/path/to/gemdir/bundler/gems/manageiq-ui-classic-2d2d5dd84fad
```

In the previous implementation of the `shared.js`, this means the pack location for application.js would end up as:

```
"manageiq-ui-classic-2d2d5dd84fad/application.js"=>"/packs/manageiq-ui-classic-2d2d5dd84fad/application.js"
```

But the views would be looking for it at:

```
/packs/manageiq-ui-classic/application.js
```

which wouldn't exist.  This changes up the method for determining the `pluginName` by including it as part of key in the `config/webpack/paths.json` file for each engine entry.  The current key isn't being used for anything, so this allows for `pluginName` to be generated in ruby.

Changes
-------

Before this PR, the `paths.json` would be generated to look like this:

```json
{
  "info": "This file is autogenerated by rake webpack:paths, do not edit",
  "output": "/Users/nicklamuro/code/redhat/manageiq",
  "engines": {
    "ManageIQ::UI::Classic::Engine": "/Users/nicklamuro/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-1363e52d11c2"
  }
}
```

Now, the engine line look like this:

```json
  "engines": {
    "manageiq-ui-classic": "/Users/nicklamuro/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-1363e52d11c2"
  }
```

Any of the changes to the rake task file are to support presenting the `engines` hash like this.

The `config/webpack/shared.js` has been updated so that the previous implementation of:

```javascript
const pluginName = dirname(entry).substring(0, dirname(entry).length - join(entryPath).length)
```

Is no longer needed, and `pluginName` comes from the hash key from `"engines"` above.  The `packPaths` variable in `config/webpack/shared.js`, that used to be populated to look like this:

```ruby
[
  '/Users/nicklamuro/code/redhat/manageiq-ui-classic-repo/app/javascript/packs/application.js',
  '/Users/nicklamuro/code/redhat/manageiq-ui-classic-repo/app/javascript/packs/hello_angular.js'
]
```

Will now look like this:

```ruby
{ 
  'manageiq-ui-classic/application': [
    '/Users/nicklamuro/code/redhat/manageiq-ui-classic-repo/app/javascript/packs/application.js',
    '/Users/nicklamuro/code/redhat/manageiq-ui-classic-repo/app/javascript/packs/hello_angular.js'
  ]
}
```

Allowing the reduce when assigning `entry:` to work and produce:

```ruby
{
  entry: {
    'manageiq-ui-classic/application': '/Users/nicklamuro/code/redhat/manageiq-ui-classic-repo/app/javascript/packs/application.js',
    'manageiq-ui-classic/hello_angular': '/Users/nicklamuro/code/redhat/manageiq-ui-classic-repo/app/javascript/packs/hello_angular.js' 
  }
}
```

### `manifest.json`

**BEFORE**

```json
{
  "manageiq-ui-classic-1363e52d11c2/application.js": "/packs/manageiq-ui-classic-1363e52d11c2/application.js",
  "manageiq-ui-classic-1363e52d11c2/hello_angular.js": "/packs/manageiq-ui-classic-1363e52d11c2/hello_angular.js"
}
```

**AFTER**

```json
{
  "manageiq-ui-classic/application.js": "/packs/manageiq-ui-classic/application.js",
  "manageiq-ui-classic/hello_angular.js": "/packs/manageiq-ui-classic/hello_angular.js"
}
```


Links
-----

* Bug introduced:  https://github.com/ManageIQ/manageiq-ui-classic/pull/1919
* Reported here:  https://gitter.im/ManageIQ/manageiq/ui?at=59936cf1ee5c9a4c5fc3f136


Steps for Testing/QA
--------------------

To confirm the failure:

1. Either
    - make sure you don't have have an `override_gem` for `manageiq-ui-classic` (better to do the next one as well, though, as it is needed to confirm the fix)
    - move the `manageiq-ui-classic` folder name to something like `manageiq-ui-classic-local` (make sure to update your entry of `override_gem` to reflect this new path)
2. From `ManageIQ/manageiq:master` up to date
3. Run `bin/update`
4. Run `rails s`
5. Try logging in

An error will happen after login.


To confirm the fix, check out this branch on your local copy of `manageiq-ui-classic`, make sure `override_gem` is pointing to it, and run steps 2 through 5 from above
